internal/blockchain/ethereum/ethereum.go

```diff
@@ -173,79 +178,131 @@ func (e *Ethereum) Init(ctx context.Context, config config.Section, callbacks bl
-		return err
-	}
 
 	e.closed = make(chan struct{})
 	go e.eventLoop()
 
 	return nil
 }
 
-func (e *Ethereum) Start() error {
+func (e *Ethereum) Start() (err error) {
 	return e.wsconn.Connect()
 }
 
+func (e *Ethereum) resolveFireFlyContract(ctx context.Context, contractIndex int) (address, fromBlock string, err error) {
+
+	if e.contractConfSize > 0 || contractIndex > 0 {
+		// New config (array of objects under "fireflyContract")
+		if contractIndex >= e.contractConfSize {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgInvalidFireFlyContractIndex, fmt.Sprintf("blockchain.ethereum.fireflyContract[%d]", contractIndex))
+		}
+		entry := e.contractConf.ArrayEntry(contractIndex)
+		address = entry.GetString(FireFlyContractAddress)
+		if address == "" {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "address", "blockchain.ethereum.fireflyContract")
+		}
+		fromBlock = entry.GetString(FireFlyContractFromBlock)
+	} else {
+		// Old config (attributes under "ethconnect")
+		address = e.ethconnectConf.GetString(EthconnectConfigInstanceDeprecated)
+		if address != "" {
+			log.L(ctx).Warnf("The %s.%s config key has been deprecated. Please use %s.%s instead",
+				EthconnectConfigKey, EthconnectConfigInstanceDeprecated,
+				FireFlyContractConfigKey, FireFlyContractAddress)
+		} else {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "instance", "blockchain.ethereum.ethconnect")
+		}
+		fromBlock = e.ethconnectConf.GetString(EthconnectConfigFromBlockDeprecated)
+		if fromBlock != "" {
+			log.L(ctx).Warnf("The %s.%s config key has been deprecated. Please use %s.%s instead",
+				EthconnectConfigKey, EthconnectConfigFromBlockDeprecated,
+				FireFlyContractConfigKey, FireFlyContractFromBlock)
+		}
+	}
+
+	// Backwards compatibility from when instance path was not a contract address
+	if strings.HasPrefix(strings.ToLower(address), "/contracts/") {
+		address, err = e.getContractAddress(ctx, address)
+		if err != nil {
+			return "", "", err
+		}
+	} else if strings.HasPrefix(address, "/instances/") {
+		address = strings.Replace(address, "/instances/", "", 1)
+	}
+
+	address, err = validateEthAddress(ctx, address)
+	return address, fromBlock, err
+}
+
+func (e *Ethereum) ConfigureContract(ctx context.Context, contracts *core.FireFlyContracts) (err error) {
+
+	log.L(ctx).Infof("Resolving FireFly contract at index %d", contracts.Active.Index)
+	address, fromBlock, err := e.resolveFireFlyContract(ctx, contracts.Active.Index)
+	if err != nil {
+		return err
+	}
+
+	e.fireflyContract = address
+	e.fireflyFromBlock = fromBlock
+	e.initInfo.sub, err = e.streams.ensureFireFlySubscription(ctx, e.fireflyContract, e.fireflyFromBlock, e.initInfo.stream.ID, batchPinEventABI)
+	if err == nil {
+		contracts.Active.Info = fftypes.JSONObject{
+			"address":      address,
+			"fromBlock":    fromBlock,
+			"subscription": e.initInfo.sub.ID,
+		}
```

Comment:
Note that there's no validation of the current config values against the contracts info that was passed in from the database - the config always overwrites what is there (and ultimately gets stored back in the database by the caller). I toyed briefly with trying to flag errors for "invalid changes" in your local config, but that got sticky.